tsa/client.go

```diff
@@ -0,0 +1,551 @@
+	// balancer so validating it gets a bit more complicated
+
+	for _, key := range client.HostKeys {
+		if key.Type() == remoteKey.Type() && bytes.Equal(key.Marshal(), remoteKey.Marshal()) {
+			return nil
+		}
+	}
+
+	return errors.New("remote host public key mismatch")
+}
+
+func (client *Client) keepAlive(ctx context.Context, sshClient *ssh.Client, tcpConn *net.TCPConn) {
+	logger := lagerctx.WithSession(ctx, "keepalive")
+
+	kas := time.NewTicker(5 * time.Second)
+
+	for {
+		// ignore reply; server may just not have handled it, since there's no
+		// standard keepalive request name
+
+		_, _, err := sshClient.Conn.SendRequest("keepalive", true, []byte("sup"))
+		if err != nil {
+			logger.Error("failed", err)
+			return
+		}
+
+		select {
+		case <-kas.C:
+			logger.Debug("tick")
+
+		case <-ctx.Done():
+			logger.Debug("stopping")
+
+			if err := tcpConn.SetKeepAlive(false); err != nil {
+				logger.Error("failed-to-disable-keepalive", err)
+				return
+			}
+
+			return
+		}
+	}
+}
+
+func (client *Client) run(ctx context.Context, sshClient *ssh.Client, command string, stdout io.Writer) error {
+	argv := strings.Split(command, " ")
+	commandName := ""
+	if len(argv) > 0 {
+		commandName = argv[0]
+	}
+
+	logger := lagerctx.WithSession(ctx, "run", lager.Data{
+		"command": commandName,
+	})
+
+	sess, err := sshClient.NewSession()
+	if err != nil {
+		logger.Error("failed-to-open-session", err)
+		return err
+	}
+
+	defer sess.Close()
+
+	workerPayload, err := json.Marshal(client.Worker)
+	if err != nil {
+		return err
+	}
+
+	sess.Stdin = bytes.NewBuffer(workerPayload)
+	sess.Stdout = stdout
+	sess.Stderr = os.Stderr
+
+	err = sess.Start(command)
+	if err != nil {
+		logger.Error("failed-to-start-command", err)
+		return err
+	}
+
+	errs := make(chan error, 1)
+	go func() {
+		errs <- sess.Wait()
+	}()
+
+	select {
+	case <-ctx.Done():
+		logger.Info("context-done", lager.Data{
+			"context-error": ctx.Err(),
+		})
+
+		err := sess.Signal(ssh.SIGINT)
+		if err != nil {
+			logger.Error("failed-to-send-signal", err)
+			return err
+		}
+
+		logger.Info("signal-sent")
+
+		err = <-errs
+		if err != nil {
+			logger.Error("command-exited-after-signal", err)
+		} else {
+			logger.Debug("command-exited-after-signal")
+		}
+
+		return err
+	case err := <-errs:
+		if err != nil {
+			logger.Error("command-failed", err)
+			return err
+		}
+
+		logger.Debug("command-exited")
+		return nil
+	}
+}
+
+func proxyListenerTo(ctx context.Context, listener net.Listener, network string, addr string) {
+	for {
+		rConn, err := listener.Accept()
+		if err != nil {
+			break
+		}
+
+		go handleForwardedConn(ctx, rConn, network, addr)
+	}
+}
+
+func handleForwardedConn(ctx context.Context, rConn net.Conn, network string, addr string) {
+	logger := lagerctx.WithSession(ctx, "forward-conn", lager.Data{
+		"network": network,
+		"addr":    addr,
+	})
+
+	defer rConn.Close()
+
+	var lConn net.Conn
```

Comment:
I'm assuming the "l" in `lConn` is "local"? Can we call it `localConn` to clarify that? Same for `rConn`.

Reply:
sure